### ISSUE TITLE
Stack traces in debug mode point to JavaScript instead of Dart

### ISSUE THREAD

**Issue by yjbanov** [flutter/flutter#50753]
This started happening as a result of the following two changes:

- https://github.com/flutter/flutter/pull/50365 - this was the first change that broke debug stack traces: the breakage was fatal, we weren't able to `toString` stack traces at all.
- https://github.com/flutter/flutter/pull/50702 - partially fixes the problem: we can `toString` stack traces, but the output of `toString` is incorrect. More details below.

Currently the stack traces point JavaScript code instead of Dart. Example:

```
Expected:
StackFrame(#-1, package:web_integration/stack_trace.dart:120:3, className: <unknown>, method: baz)
StackFrame(#-1, package:web_integration/stack_trace.dart:115:3, className: <unknown>, method: bar)
StackFrame(#-1, package:web_integration/stack_trace.dart:110:3, className: <unknown>, method: foo)
Actual:
StackFrame(#-1, package:web_integration/stack_trace.dart.lib.js:138:15, className: <unknown>, method: baz)
StackFrame(#-1, package:web_integration/stack_trace.dart.lib.js:135:17, className: <unknown>, method: bar)
StackFrame(#-1, package:web_integration/stack_trace.dart.lib.js:132:17, className: <unknown>, method: foo)
```

Temporarily I will add a test that matches current behavior (see https://github.com/flutter/flutter/pull/50652), but I'll leave a TODO to change the test.

/cc @jonahwilliams

**Comment by jonahwilliams** [comment]
The issue currently is that all of trackLibraries calls used to map module name -> source map are null:

```
  dart.defineLazy(actions$.SelectAction, {
    /*actions$.SelectAction.key*/get key() {
      return C1158 || CT.C1158;
    }
  });
  dart.trackLibraries("null", {
    "package:flutter/src/widgets/widget_span.dart": widget_span,
    "package:flutter/src/widgets/table.dart": table$,
    "package:flutter/src/widgets/inherited_model.dart": inherited_model,
    "package:flutter/src/widgets/widget_inspector.dart": widget_inspector,
    "package:flutter/src/widgets/title.dart": title$,
     ...
}
```

This is likely due to a misconfiguration with or in the frontend_server

**Comment by jonahwilliams** [comment]
Even with a very simple configuration, I can repro track libraries as null issue:

```
/Users/jonahwilliams/Documents/flutter/bin/cache/dart-sdk/bin/dart /Users/jonahwilliams/Documents/flutter/bin/cache/artifacts/engine/darwin-x64/frontend_server.dart.snapshot --sdk-root
/Users/jonahwilliams/Documents/flutter/bin/cache/flutter_web_sdk/ --incremental --target=dartdevc -Ddart.developer.causal_async_stacks=true --output-dill
/var/folders/sq/vd7vtp0j60ld19wt58vg7n5c00c4f_/T/flutter_tool.nyPVEO/app.dill --packages /Users/jonahwilliams/Documents/flutter/examples/hello_world/.packages -Ddart.vm.profile=false
-Ddart.vm.product=false --bytecode-options=source-positions,local-var-info,debugger-stops,instance-field-initializers,keep-unreachable-code,avoid-closure-call-instructions --enable-asserts
--track-widget-creation --filesystem-scheme org-dartlang-root --platform file:///Users/jonahwilliams/Documents/flutter/bin/cache/flutter_web_sdk/kernel/flutter_ddc_sdk.dill
```

**Comment by jonahwilliams** [comment]
fyi @jakemac53

**Comment by jonahwilliams** [comment]
(There is a second issue that the bootstrap script is mapping file name to module name, but that is quite minor). The fix for this locally is:

```
@@ -194,7 +194,13 @@ define("main_module.bootstrap", ["$entrypoint", "dart_sdk"], function(app, dart_
       function(url) {
         url = url.replace(baseUrl, '/');
         var module = window.\$dartLoader.urlToModuleId.get(url);
-        if (!module) return null;
+        if (!module) {
+          if (url.endsWith('dart_sdk.js')) {
+            module = 'dart_sdk';
+          } else {
+            module = 'packages' + url.split('packages')[1].split('.')[0];
+          }
+        }
         return dart.getSourceMap(module);
```

**Comment by grouma** [comment]
Note that the bootstrap logic has been updated and pushed into package:dwds - see https://github.com/dart-lang/webdev/pull/898

**Comment by yjbanov** [comment]
Can we assign someone to this issue? It's pretty serious as you cannot debug an app in Dart any more.

BTW, when the fix rolls in, it will, expectedly, break this test: https://github.com/flutter/flutter/blob/master/dev/integration_tests/web/lib/stack_trace.dart. The test will need to be updated to reflect the correct stack trace format.

/cc @vsmenon

**Comment by jonahwilliams** [comment]
I think this is currently fixed in dart, and requires a minor change in the tool when the roll happens. I'll take care of the later part

**Comment by jonahwilliams** [comment]
The roll shouldn't break the test, because there is a required secondary tool change needed

**Comment by jonahwilliams** [comment]
The fix for this has rolled into the engine, currently waiting on a framework roll

**Comment by MatthewPatience** [comment]
Has this made it's way into production yet? I still am unable to view references to dart files in the stack traces for web. I'm on Flutter 2.2.1

**Comment by maxltucker** [comment]
> 
> 
> Has this made it's way into production yet? I still am unable to view references to dart files in the stack traces for web. I'm on Flutter 2.2.1

I recently upgraded to 2.2.1 and had this issue pop up. Perhaps it was reintroduced? My stack traces point to .js files, but only if the error is thrown from inside an event listener.

**Comment by Soorya-S** [comment]
Does anyone have a solution for this?
this issue still exists on latest flutter version 

Flutter 2.4.0-0.0.pre • channel dev • https://github.com/flutter/flutter.git
Framework • revision 96bbcd006f (12 days ago) • 2021-06-30 03:43:41 -0400
Engine • revision c633b2a6a0
Tools • Dart 2.14.0 (build 2.14.0-258.0.dev)

This issue causes lot of negative effect in productivity & I'm thinking about switching back to js framework. 
Flutter team should take this issue seriously.

**Comment by vsmenon** [comment]
Please open a new issue with repros.  This is very likely to be a different underlying problem.

FWIW, if I run the counter sample from VSCode on Chrome and add an exception to `_incrementCounter`, I see a mapped stack trace (tip-of-tree flutter build) in the VSCode debug console:

```
════════ Exception caught by gesture ═══════════════════════════════════════════
The following message was thrown while handling a gesture:
An error

When the exception was thrown, this was the stack
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 251:49  throw_
packages/counter/main.dart 59:7                                               <fn>
packages/flutter/src/widgets/framework.dart 1088:30                           setState
packages/counter/main.dart 52:5                                               [_incrementCounter]
packages/flutter/src/material/ink_well.dart 989:21                            [_handleTap]
packages/flutter/src/gestures/recognizer.dart 193:24                          invokeCallback
packages/flutter/src/gestures/tap.dart 608:48                                 handleTapUp
packages/flutter/src/gestures/tap.dart 296:5                                  [_checkUp]
packages/flutter/src/gestures/tap.dart 267:7                                  acceptGesture
packages/flutter/src/gestures/arena.dart 157:12                               sweep
packages/flutter/src/gestures/binding.dart 444:20                             handleEvent
packages/flutter/src/gestures/binding.dart 420:14                             dispatchEvent
packages/flutter/src/rendering/binding.dart 289:11                            dispatchEvent
packages/flutter/src/gestures/binding.dart 374:7                              [_handlePointerEventImmediately]
packages/flutter/src/gestures/binding.dart 338:5                              handlePointerEvent
packages/flutter/src/gestures/binding.dart 296:7                              [_flushPointerEventQueue]
packages/flutter/src/gestures/binding.dart 279:32                             [_handlePointerDataPacket]
lib/_engine/engine/platform_dispatcher.dart 989:13                            invoke1
lib/_engine/engine/platform_dispatcher.dart 176:5                             invokeOnPointerDataPacket
lib/_engine/engine/pointer_binding.dart 129:39                                [_onPointerData]
lib/_engine/engine/pointer_binding.dart 554:18                                <fn>
lib/_engine/engine/pointer_binding.dart 507:21                                <fn>
lib/_engine/engine/pointer_binding.dart 215:16                                <fn>
Handler: "onTap"
Recognizer: TapGestureRecognizer#11730
    debugOwner: GestureDetector
    state: ready
    won arena
    finalPosition: Offset(1749.2, 872.1)
    finalLocalPosition: Offset(25.2, 42.1)
    button: 1
    sent tap down
════════════════════════════════════════════════════════════════════════════════

```
